apps/omg_watcher/lib/omg_watcher/exit_processor_dispatcher.ex

```diff
@@ -0,0 +1,274 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.Watcher.ExitProcessorDispatcher do
+  @moduledoc """
+  Pretty sure it's doing some cool stuff.
+  """
+
+  require Logger
+
+  @doc """
+  Checks validity of all exit-related events and returns the list of actionable items.
+  Works with `OMG.State` to discern validity.
+
+  This function may also update some internal caches to make subsequent calls not redo the work,
+  but under unchanged conditions, it should have unchanged behavior from POV of an outside caller.
+  """
+  def check_validity() do
+    do_chack_validity()
+  end
+
+  def check_validity(timeout) do
+    do_chack_validity(timeout)
+  end
+
+  defp do_chack_validity(timeout \\ 5000) do
+    forwarded_result = forward(:check_validity, timeout)
+
+    Enum.reduce(forwarded_result, {:ok, []}, fn {chain_validity, events}, acc ->
+      {acc_chain_validity, acc_events} = acc
+
+      case acc_chain_validity do
+        :ok ->
+          {chain_validity, acc_events ++ events}
+
+        {:error, :unchallenged_exit} ->
+          {{:error, :unchallenged_exit}, acc_events ++ events}
+      end
+    end)
+  end
+
+  @doc """
+  Returns all information required to produce a transaction to the root chain contract to present a competitor for
+  a non-canonical in-flight exit
+  """
+  def get_competitor_for_ife(txbytes) do
+    forward_single_result({:get_competitor_for_ife, txbytes})
+  end
+
+  @doc """
+  Returns all information required to produce a transaction to the root chain contract to present a proof of canonicity
+  for a challenged in-flight exit
+  """
+  def prove_canonical_for_ife(txbytes) do
+    forward_single_result({:prove_canonical_for_ife, txbytes})
+  end
+
+  @doc """
+  Returns all information required to challenge an invalid input piggyback
+  """
+  def get_input_challenge_data(txbytes, input_index) do
+    forward_single_result({:get_input_challenge_data, txbytes, input_index})
+  end
+
+  @doc """
+  Returns all information required to challenge an invalid output piggyback
+  """
+  def get_output_challenge_data(txbytes, output_index) do
+    forward_single_result({:get_output_challenge_data, txbytes, output_index})
+  end
+
+  @doc """
+  Returns challenge for an invalid standard exit
+  """
+  def create_challenge(exiting_utxo_pos) do
+    forward_single_result({:create_challenge, exiting_utxo_pos})
+  end
+
+  @doc """
+  Returns a map of requested in flight exits, keyed by transaction hash
+  """
+  def get_active_in_flight_exits() do
+    in_flight_exits =
+      :get_active_in_flight_exits
+      |> forward()
+      |> Enum.reduce([], fn {:ok, in_flight_exits}, acc ->
+        acc ++ in_flight_exits
+      end)
+
+    {:ok, in_flight_exits}
+  end
+
+  # TODO: move private funcs at the bottom of the file
+  defp forward(func) do
+    Enum.map(OMG.WireFormatTypes.exit_game_tx_types(), fn tx_type ->
+      GenServer.call(tx_type, func)
```

Comment:
it's pretty expensive to sync call in a tight loop. might be better to decouple these.